validator/client/src/main/java/tech/pegasys/teku/validator/client/StableSubnetSubscriber.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.validator.client;
+
+import static java.lang.Integer.min;
+import static tech.pegasys.teku.util.config.Constants.ATTESTATION_SUBNET_COUNT;
+import static tech.pegasys.teku.util.config.Constants.EPOCHS_PER_RANDOM_SUBNET_SUBSCRIPTION;
+import static tech.pegasys.teku.util.config.Constants.RANDOM_SUBNETS_PER_VALIDATOR;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.NavigableSet;
+import java.util.Optional;
+import java.util.Random;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import tech.pegasys.teku.validator.api.SubnetSubscription;
+import tech.pegasys.teku.validator.api.ValidatorApiChannel;
+
+public class StableSubnetSubscriber {
+
+  private final ValidatorApiChannel validatorApiChannel;
+  private final Set<Integer> availableSubnetIndices =
+      IntStream.range(0, ATTESTATION_SUBNET_COUNT).boxed().collect(Collectors.toSet());
+  private final NavigableSet<SubnetSubscription> subnetSubscriptions =
+      new TreeSet<>(Comparator.comparing(SubnetSubscription::getUnsubscriptionSlot));
+  private final Random rand = new Random();
+
+  private volatile int validatorCount;
+
+  public StableSubnetSubscriber(ValidatorApiChannel validatorApiChannel, int validatorCount) {
+    this.validatorApiChannel = validatorApiChannel;
+    this.validatorCount = validatorCount;
+    onSlot(UnsignedLong.ZERO);
```

Comment:
I don't think we should call this on creation - it will wind up activating things during the construction phase before we finish wiring everything in.  I'd probably just wait for the first actual slot event.

Reply:
Fair point. I was undecided.

Reply:
Done.